Swc 130 implementar web socket para mostrar la carta de movimiento

MatiasScantamburlo opened the pull request:
¡Chicos acá dejo una PR! Espero sus comentarios :100:

MatiasScantamburlo committed:
Se crea endpoint PUT /api/lobby/{game_id}/select

- Permite que un jugador seleccione una carta en su turno de juego
- Verifica que el juego, jugador y carta existan en sus respectivos
    repositorios.
- Asegura que el jugador esté participando en el juego y que sea su turno.
- Usa el "Manager" para transmitir a todos los jugadores del lobby la
    selección de una carta, junto con el nombre e ID del jugador y de la carta.
- Retorna un mensaje de éxito en caso de que todas las validaciones sean
    superadas.

 En la rama SWC-130-Implementar-WebSocket-para-mostrar-la-carta-de-movimiento

 Cambios a ser confirmados:
	modificados:     main.py

MatiasScantamburlo committed:
Se agrega test para el endpoint PUT /api/lobby/{game_id}/select

- El test verifica la validación de jugadores y cartas, la correcta selección
    de la carta y el broadcast adecuado, de momento a si mismo.
- Repositorios mockeados (GameRepository, PlayerRepository, FigRepository)
    para los datos de juego, jugador y carta.
- Se probaron conexiones WebSocket para asegurar que el mensaje de broadcast
    se envíe correctamente tras la selección de carta.
- Format en test_notify_lobby

TODO: Verificar para demas conexiones.

 En la rama SWC-130-Implementar-WebSocket-para-mostrar-la-carta-de-movimiento

 Cambios a ser confirmados:
	modificados:     tests/test_notify_lobby.py
	nuevos archivos: tests/test_select_card.py

MatiasScantamburlo committed:
Se verifica el broadcast en múltiples websockets tras selección de carta

- Añadido test para verificar que el mensaje de broadcast se reciba en
    múltiples WebSockets cuando un jugador selecciona una carta.
- El test ahora comprueba que el mismo mensaje de selección de carta se envíe
    y reciba correctamente en los otros jugadores conectados al lobby.

 En la rama SWC-130-Implementar-WebSocket-para-mostrar-la-carta-de-movimiento

 Cambios a ser confirmados:
	modificados:     tests/test_select_card.py

MatiasScantamburlo committed:
Merge branch 'develop' into SWC-130-Implementar-WebSocket-para-mostrar-la-carta-de-movimiento

MatiasScantamburlo committed:
Se cambia PUT por WS

- Por simplicidad se crea un ws que realice las funciones del endpoint
- Se corrige al acceso a cartas de movimiento

En la rama SWC-130-Implementar-WebSocket-para-mostrar-la-carta-de-movimiento

 Cambios a ser confirmados:
	modificados:     main.py

MatiasScantamburlo committed:
cambios para después hacer un ticket independiente

MatiasScantamburlo committed:
Completamos test correspondiente a /ws/lobby/{game_id}/select

- Se crea logica de testeo al rededor de PlayerInfo para poder
      corroborar las cartas en mano.
- Se verifica que el mensaje enviado sea recivido por los jugadores.
- Se agrega manejador CARDS en services/connection_manager
- Se realiza un format en main, se limpian los prints
- Se agrega descripcion al ws implementado

En la rama SWC-130-Implementar-WebSocket-para-mostrar-la-carta-de-movimiento

 Cambios a ser confirmados:
	modificados:     main.py
	modificados:     model/__init__.py
	modificados:     services/connection_manager.py
	modificados:     tests/test_select_card.py

MatiasScantamburlo committed:
Merge branch 'develop' into SWC-130-Implementar-WebSocket-para-mostrar-la-carta-de-movimiento

MatiasScantamburlo committed:
Se agrega identifier

- Se utiliza el campo identifier para identificar a que jugador
    pertenece la figura de movimiento seleccionada.
- Se condiciona test para esto.

En la rama SWC-130-Implementar-WebSocket-para-mostrar-la-carta-de-movimiento

 Cambios a ser confirmados:
	modificados:     main.py
	modificados:     tests/test_select_card.py

MatiasScantamburlo committed:
Se arregla comentario

KaiserOfNone commented on main.py:
```diff
@@ -544,6 +541,14 @@ async def lobby_notify_inout(websocket: WebSocket, game_id: int, player_id: int)
+    if game is None:
+        raise HTTPException(status_code=404, detail="Game not found")
+
+    player = player_repo.get(player_id)
```
player lo estas obteniendo en la linea 556 abajo.

MatiasScantamburlo replied:
La primera vez tiene sentido para corroborar que exista el player antes de agregar su conexion al diccionario. Lo que podría estar en duda es si necesitamos que se pase el identifier, ya que seguramente será el del jugador previamente conectado, eso se podría mejorar.

Sin embargo corroborar que el identifier pasado existe, por ende calcular 'player', no está mal.

KaiserOfNone commented on main.py:
```diff
@@ -559,10 +564,8 @@ async def lobby_notify_inout(websocket: WebSocket, game_id: int, player_id: int)
-            if game is None:
-                await websocket.send_json({"error": "Game not found"})
-                continue
+            game.add_player(player)
```
El ws no es responsable de agregar jugadores, para eso esta el endpoint.

seedLou replied:
El ws tambien puede agregar al jugador (que era la unica responsabilidad del endpoint ademas de las verificaciones que el ws por cierto tambien hace)  funciono bien en la integracion, por eso dijimos ¿para que queremos al endpoint si lo unico que hace tambien puede hacerlo el ws?

KaiserOfNone commented on main.py:
```diff
@@ -590,3 +596,52 @@ async def lobby_notify_status(websocket: WebSocket, game_id: int, player_id: int
+    """
+    Este ws se encarga de recibir la selección de cartas de un jugador y notificar a los demás jugadores de la partida.
+
+    Se espera: {card_id: 'valor', player_identifier: 'valor'}
```
Pone tipos, no solo valor, eso no me dice nada.

MatiasScantamburlo replied:
Voy a mejorar los comentarios remplazando 'valor' por el tipo que se espera. Por ejemplo:

{card_id: 'int', player_identifier: 'str'}

KaiserOfNone commented on main.py:
```diff
@@ -590,3 +596,52 @@ async def lobby_notify_status(websocket: WebSocket, game_id: int, player_id: int
+                await websocket.send_json({"error": "Player not in game"})
+                continue
+
+            hand = game.player_info[current_player.id].hand_mov
```
Esto es horrible. Usa un metodo en game para sacar esta data, player_info es un detalle de implementacion.

seedLou replied:
Me sirve porque tambien use descaradamente player_info 
Ahora creo un metodo para acceder a esa data sin romper toda la abstraccion ahhhhh curioso

MatiasScantamburlo replied:
Banco, mejoremos esto

KaiserOfNone commented on main.py:
```diff
@@ -590,3 +596,52 @@ async def lobby_notify_status(websocket: WebSocket, game_id: int, player_id: int
+    if game is None:
+        raise HTTPException(status_code=404, detail="Game not found")
+
+    player = player_repo.get(player_id)
+    if player is None:
+        raise HTTPException(status_code=404, detail="Player not found")
```
Esto no hace falta. El primer mensaje te debe dar el identifier de todas formas, podes extraer el id de ahi,

MatiasScantamburlo replied:
¿Porqué decís que no hace falta? ¿Como corroboro que no se conecten con un game_id inexistente? Recordá que desde el último cambio de la Lou, el método connect pide el player_id, si no me conecto primero no puedo escuchar el identifier.

KaiserOfNone commented on model/__init__.py:
```diff
@@ -1,5 +1,5 @@
 from .board import SIZE_BOARD
 from .fig_cards import FigCards
-from .game import Game
+from .game import Game, PlayerInfo
```
No exportes PlayerInfo, no lo exporte a proposito.

MatiasScantamburlo replied:
Lo necesito para crear instancias de figura en la mano del jugador. Puedo crear un método que abstraiga esto si no querés importar PlayerInfo

MatiasScantamburlo replied:
Ahí lo dejé mejor y ya no hay necesidad de importarlo :heart_hands:

KaiserOfNone commented on tests/test_select_card.py:
```diff
@@ -0,0 +1,92 @@
+            self.player_repo.save(p)
+
+        self.game = Game(
+            id=101010,
```
No hardcodees la ID, genera tests que se pueden romper facilmente. Cuando haces save los campos con valores por defecto se llenan automaticamente.

MatiasScantamburlo replied:
¡Listo!

KaiserOfNone commented on tests/test_select_card.py:
```diff
@@ -0,0 +1,92 @@
+
+            # agregamos cartas en la manos de los jugadores en juego
+            for idx, player in enumerate(self.game.players, start=1):
+                player_info = PlayerInfo(
+                    player_id=player.id,
+                    turn_position=idx,
+                    hand_mov=[idx + 1, idx + 2, idx + 3],
+                )
```
No hagas esto, sobreescribi los campos que necesitas, o mejor todavia, agrega metodos para trabajar la data que necesitas.

MatiasScantamburlo replied:
¡Hecho!

KaiserOfNone commented on tests/test_select_card.py:
```diff
@@ -0,0 +1,92 @@
+                self.game.player_info[player.id] = player_info
+
+            # conectamos ambos jugadores
+            with self.client.websocket_connect(
+                f"/ws/lobby/101010/select?player_id={player0.id}"
+            ) as websocket0:
+                with self.client.websocket_connect(
+                    f"/ws/lobby/101010/select?player_id={player1.id}"
+                ) as websocket1:
```
Pone todas las conecciones al ws al principio, en otros tests lo estamos haciendo asi.

MatiasScantamburlo replied:
Mi vida por la horda

KaiserOfNone commented on main.py:
```diff
@@ -590,3 +596,52 @@ async def lobby_notify_status(websocket: WebSocket, game_id: int, player_id: int
+    """
+    Este ws se encarga de recibir la selección de cartas de un jugador y notificar a los demás jugadores de la partida.
+
+    Se espera: {card_id: 'valor', player_identifier: 'valor'}
```
Agrega un campo al json que se retorna que sea `"action": "play"`, lo necesitamos para cuando tengamos el `"action": "undo"` para saber cuando mostrar y esconder cartas en el front.

seedLou replied:
Todas las cartas de movimiento estan ocultas hasta que se juegan. Si los demas jugadores reciben el json con el player_id de otro jugador entonces la carta se muestra hasta que se descarte (ya no existe in deck) o se reviertan los movimientos parciales (aca el action seria "undo" nuevamente)

javierscantamburlo-source committed:
Fixes

- Mejoramos Docstrings.
- Creamos en model/game current_hand_mov() y add_hand_mov().
- No se importa PlayerInfo desde game.
- Se quita id de game, ws se abre al inicio, se toman cartas aleatoreas.

 On branch SWC-130-Implementar-WebSocket-para-mostrar-la-carta-de-movimiento

 Changes to be committed:
	modified:   main.py
	modified:   model/__init__.py
	modified:   model/game.py
	modified:   tests/test_select_card.py

javierscantamburlo-source committed:
Merge branch 'develop' into SWC-130-Implementar-WebSocket-para-mostrar-la-carta-de-movimiento
